src/Orleans/Messaging/IncomingMessageBuffer.cs

```diff
@@ -154,7 +161,11 @@ public bool TryDecodeMessage(out Message msg)
             List<ArraySegment<byte>> body = ByteArrayBuilder.BuildSegmentListWithLengthLimit(readBuffer, bodyOffset, bodyLength);
             
             // build message
-            msg = new Message(header);
+            var context = new DeserializationContext(this.serializationManager)
```

Comment:
We can reduce the allocations below the number of allocations in the existing code with a little bit of reorganization.